internal/topology/check/compatibility.go

```diff
@@ -18,39 +18,179 @@ limitations under the License.
 package check
 
 import (
-	"github.com/pkg/errors"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 // ReferencedObjectsAreStrictlyCompatible checks if two referenced objects are strictly compatible, meaning that
 // they are compatible and the name of the objects do not change.
-func ReferencedObjectsAreStrictlyCompatible(current, desired client.Object) error {
+func ReferencedObjectsAreStrictlyCompatible(current, desired client.Object) field.ErrorList {
+	var allErrs field.ErrorList
 	if current.GetName() != desired.GetName() {
-		return errors.Errorf("invalid operation: it is not possible to change the name of %s/%s from %s to %s",
-			current.GetObjectKind().GroupVersionKind(), current.GetName(), current.GetName(), desired.GetName())
+		allErrs = append(allErrs, field.Invalid(
+			field.NewPath("metadata", "name"),
+			desired.GetName(),
+			fmt.Sprintf("must be equal to %v for %v", current.GetName(), desired.GetName()),
```

Comment:
Wondering if we can provide a better error message for compatibility checks.
What about:
- `field.Forbidden` given that ErrorTypeForbidden is used to report valid (as per formatting rules) values which would be accepted under some conditions, but which are not permitted by the current conditions, which seems a better fit than `field.Invalid`.
- change the error message into something like "value cannot be set to %v, this can lead to incompatible changes when transitioning from current value %v"